# Analyzing Names from Facebook 

By using the last name lists collected previously, I was able to search for people's name on public search on Facebook.
1101 total last names were searched on FB, and total of 164075 names (valid and invalid, also including non-Japanese names) were collected. 

In order to be able to leverage this data, we'll need to clean this up first, concretely,
1. extract valid names only.
2. classify into Japanese and non-Japanese names
3. isolate first names from Japanese names, and create a list of unique first names

Then we have the unique list of first names, that can be added to the Faker library. Other names collected here can be used for the testing purpose as well.

Now Let's get started!

In [33]:
import pandas as pd
import os
import glob

cur = os.getcwd()
PATH = os.path.join(cur, "data/scraper/fb_names")
# create a list of file names, and read the whole name in as a DataFrame, for pretty display
file_names = glob.glob(os.path.join(PATH, "*_names.txt"))
frame_generator = (pd.read_csv(f, header=0, names=['name']) for f in file_names)
df = pd.concat(frame_generator, axis=0, ignore_index=True)
print("names are loaded!")

names are loaded!


In [36]:
# now the df is loaded, let's take a look!
from IPython.display import HTML, display
pd.set_option("display.max_rows", None)
display(df.sample(500))

name
144330                          小坂田 加代
107265                          利根田 一朗
162733                  Dian'ga Diageo
124631                       Shan Shan
130343                         Yi Yuan
88367                             山下 潤
138258                             片山泉
64561                      Dao Cao Van
28756                             木下裕子
161076                       田代 アクリサイン
92218                            本間 知子
136319                           足立 勝子
148980                           江口 博子
23928                          Mai Dao
40744                         野口 チョコット
65033                         Xiao Jia
157740                          中島 由美子
68351                             馬場 隆
44980                             小川洋平
54870                            西田 幸弘
138465                            片山幸一
55128                             吉川梨恋
99824                大風 (Spring Huang)
128755                           戸田 佐枝
98381                              本田明
144865                           喜来 彩香
34371                             河村美樹
77975                             川村 学
110933                            宗像恵美
26793                             吉長翔児
70509                             栗原邦子
71147                              栗原誼
11625                             西山昭彦
10438                            成田 栄樹
107511                           利根川勝明
46028                       久世 泉名 (高橋)
68579                              馬場恵
16745                        大島萌 (まめぐ)
154325                           佐々木 悟
111390                            宗像 研
155141                     Haddy Chang
16614                       大島 直 (nao)
32541                            岩本 祐子
12325                             西山 貴
43968                           Lin Ye
140344                           大川 邦彦
77148                             増田浩之
30291                            北村 美和
81675                            実森 智弘
140484                            浜田愛子
144430                           小坂田 唯
132027                       Chi Zhang
80464                         Tang Son
30183                             北村久美
84342                          Nam Dao
37024                            石橋 佳昇
4192                             吉村 吾郎
121031                     Cun Dao Dao
162547                         Dong Ta
99101                              悟守内
88791                              宇多露
22321                            平井 謙次
34215                            河村 太郎
91349                 Yi-shan Lin (夏天)
148943                            江口洋助
137769                          Liu Ze
139177                           片山 優子
101352                           川和 怜実
81358                            水岡 晃大
133290                            菅野達也
18596                              宮田元
34663                              趙秋芳
65343             Li Jia Qi (Xiao You)
125809                      遠藤 一司 (太田)
153730                         Yan Man
123799                             石田誠
158724             きしゃ じ ょぜ (Kisshang)
49375    Bán Ve Chai (No Promise Boys)
64580                          Đào Cao
113697                      音処　かなで 大倉山
20522                            横田 正教
124544                       Shan Shan
133689                           菅野 英雄
18664                             宮田聡子
35782                            萩原 早苗
62115                            武田 翔吾
144408                          小坂田 如雲
23631                            川倉 健治
70564                          Li Yuan
34570                       Zhu Wei Lu
89876                            奥口可奈子
53681                             西川　凌
55924                           姫田 あすか
52201                            龍弥 浪花
67100                            譲原百合子
112584                            塩本直広
141957                            舟坂輝也
32580                             岩本志穂
13280                            宮本明日香
91061                          Yi Shan
60856                            武田 陽子
34421                              河村賢
23678            

As you can see, most of them are valid names (Japanese and non-Japanese), but some includes () entry in which they have second name? or nick names. 
Most of Japanese names are consisted of Kanji, with or without white space between last & first name.

## EDA on names
Exploring the characteristics of the names collected here. We'll start with the size of the dataset, and dive itno further details

Here, let's remove the records with () entry.

In [38]:
# get the size
names = df[~df['name'].str.contains("(", regex=False)]
display(names.sample(100))
print("Dataset Size: ", df.shape[0])
print("Dataset without () size: ", names.shape[0])

name
148425             江口富美雄
163532              工藤幾子
62511               田中優輝
26946              牧野 時夫
163393            木ノ下 拓弥
13674               宮本貴文
30108              北村 信明
143392          Xiao Bao
63448              上田 昌宏
81315               水岡 崇
41891               柴田栄二
20846               横田会美
14722             関口 アナン
83501               神田 章
136312             足立 和也
148885              江口雅彦
397                  김다예
143854             倉敷 あみ
59605               杉本味加
64877   Cao Thi Ngoc Dao
103190             中島明日香
112437              橋本純子
102872               川上優
3997              綿屋 ライン
60435              杉本絵理奈
49793               原田大輔
40713             美藤 真奈美
10603             成田 ちひろ
94290         石原 慎介 さぬき版
12031           西山 コウタロウ
144839               喜来巧
38576               赤塩 望
31791              五十嵐達也
44404         Ye Lin Swe
47375              大石 哲大
61058             武田 あゆみ
142536           新井 ジュニア
95896              上関沙也可
135564             平野 祐一
105554              渡辺直輝
10240              成田 孝平
99818      Ping Chuan Ma
23632             川倉 健二郎
146246              前川宏子
62970               上田愛美
69521              栗原 恵子
8979            Hua Yuan
122352              伊藤克弥
92158               本間郁登
17852           Grace Gu
39761               須藤芳和
150135            氷見 明日香
146372              前川 直
97800                村上稔
5594               中西 賢治
111389             宗像 里歩
124415              東間 嶺
3806               綿屋 将孝
112796               김예지
78131              川村 哲也
50122     Jian Xin Zheng
26585               亀野善徳
132674             Lu Mu
117303              広瀬愛子
23887            Mai Dao
160515     Huo Zhong Wei
126837             森田 ロキ
101934               川上薫
91599        Yi Shan Lin
130242           Yi Yuan
19398              宮田 晶子
53331              西川 智隆
50917            Xian Yu
120748         Qi Yu Rui
29388               北村知子
92544               本間亮太
41000              掛橋登代子
160812    Zhong Shun Wei
23162               小島典子
152693              高山千尋
114184           Tấn Hội
20232              智子 横田
53017              菅原 里菜
34864                顧人遠
143587          Xiao Bao
114158          Công Tấn
83741                神田哲
19434               宮田和英
6630               中西 晃治
99169             水無瀬 まり
63880               上田哲也
117700              広瀬晃久
74899               村田優介
60832              武田 克美
46097              山中香奈美
94178               松岡弘道
3421             Kai Cao
135166             市川千恵子
37501               石橋龍太
118350              菊池大介

Dataset Size:  164044
Dataset without () size:  152425


That leaves us 152425 records, out of all 164044 records. This also contains Japanese and non-Japanese names. 

### Letterwise Analyzation

Now we can analyze the records in terms of types of letters that it contains. Considering that most Japanese names on this dataset are composed sololy of Kanji, let's check the ones that contains
- Hiragana
- Katakana
individually. 

We start with removing names that contains non-Japanese characters, for now. 

In [42]:
# regex pattern for Japanese characters, thanks to 
# https://gist.github.com/ryanmcgrath/982242 
pattern = "[\u3000-\u303F]|[\u3040-\u309F]|[\u30A0-\u30FF]|[\uFF00-\uFFEF]|[\u4E00-\u9FAF]|[\u2605-\u2606]|[\u2190-\u2195]|\u203B"
j_names = names[names['name'].str.contains(pattern, regex=True)]
display(j_names.sample(100))

name
34991         成塚幸子
47916        大石 広美
77340         青木美穂
151471       小谷田 稔
122771       伊藤 達也
13782         宮本綾恵
115073       三条恵理奈
7447        早川 奈穂子
121721        河野広一
34545         河村英子
17504        大島 大容
144800        中邑友香
96733         佐野雪江
31148        五十嵐千尋
128374       鎌田 優子
136985        更科礼子
104318        石原彩乃
88196       山下 慶一郎
5541         中西 和美
64229        稲尾 英利
151459       小谷田春奈
20800         横田周斗
133511        菅野高広
127410        中川涼子
137021        青塚智義
69911        栗原 朋宏
90716         三浦信雄
25500         樺澤 良
23125         小島 建
86070         上野晋弘
140264        片山友也
89511     宇多良 カナリア
57613        渡部 一夫
30015       北村 真紀子
128548        望月大輔
8356         早川 寿朗
90534         三浦彰太
118325       菊池 良和
28394        山根 朋子
102318        川上順一
88204         山下勝彦
138278        片山幸司
78885         加藤英明
55243        吉川 知里
53479        西川 沙紀
134842        市川育子
122384         伊藤舞
44897       小川 トオル
58547          渡部茜
7865           早川泉
5160         安藤 晶子
163849      工藤 ケラネ
96931         佐野克明
118399  菊池 雄一、きくりん
76181        桜井 浩和
138543        片山 静
31131        五十嵐祐介
50371         兼松信雄
10336        成田 雄亮
150296       智貴氷見山
94919        及川 史乃
161263        田代一洋
104579       石原 直也
30675        五十嵐貴子
112275       橋本加奈子
108406        永田 静
138326       片山寄世史
61095        武田 雄一
56901         渡部陽子
137146        翔 青塚
70265         栗原 良
137025         青塚一
16782          大島萌
135176        市川恭唯
94273         深坂弘光
109260        吉田美穂
66762         斉藤裕之
110110        砂永拓馬
23532        小島 繁樹
50238         鈴木兼信
85213        安田 憲保
17694         大島春菜
161073       田代 知子
41375         小田桃子
6231          中西文彦
29452         北村直樹
112179      橋本　 純子
63769        上田 静香
145515       前川 成美
71261       栗原 かのん
126777        森田大河
30190        北村 和人
105201        腰越志穂
145561        前川茂男
10307         成田茂樹
11116        西山 洋平
72975         小松正幸
121308        村上道江
53694         西川依里
3941           吳竣豪

Now, let's take a look at names that include Hiragana and Katakana.

In [44]:
## Set up regex
hiragana = "[\u3040-\u309F]"
katakana = "[\u30A0-\u30FF]"

hira_names = names[names['name'].str.contains(hiragana, regex=True)]
display(hira_names.sample(100))

name
133448             菅野 えいじ
11043              西山 あかね
133876              森下 りお
51507            建石 せいじろう
26192              男澤 あけみ
53192              菅原 ゆう子
25786              大塚 りえこ
103005              中島 かよ
59042              渡部 めぐみ
63051              上田 まゆみ
74916              村田 ほのか
79196              高田 さやか
136382             多良間 かほ
150918             秋山 はるか
140651             浜田 めぐみ
25758               大塚 ゆみ
56601              渡部 かおり
62467              田中 みなみ
138617             片山 ゆかり
156066             貞島 みちこ
120974             道村 こうた
30549              五十嵐 ちえ
55558              吉川 さをり
124420             東間 さとる
103923             石原 さとみ
20206               横田 ゆい
100632              岩崎 かほ
140397             浜田 けい子
145630             前川 あづさ
3927          麵屋 やまと バンコク
27711              山根 ひとみ
111235             宗像 ひかる
118769         菊池 生真 釣りバカ
123622             石田 みつえ
157407             中嶋 ありす
141185             箱石 みず穂
138793             片山 あゆみ
152303             高山 かおり
152476             高山 みどり
127966             下浜 あやこ
115049            三条 はるみ屋
91950              本間 ひとみ
102214              川上 りか
48041              大石 さゆり
108837             平山 しのぶ
3771                村形 しほ
88212              山下 ひろみ
58700              渡部 にっき
116496              薮崎 まゆ
156299             永井 あすか
101651               つと 坊
59472               杉本 ゆか
123425             光 の真 実
133187             菅野 ひさ子
100161             岩崎 まゆみ
34975               成塚 あい
56029              姫田 みゆき
7317            よしはる 仁後良治
77029              増田 さつき
99353              平田 めぐみ
26544             小永井 かおり
62162             武田 しんぺー
54640              橋倉 なつき
35971              萩原 さくら
75811             桜井 ひまわり
21403              山田 みなみ
23671              岡部 ひとみ
150718              秋山 みゆ
60497              杉本 よし美
39412              須藤 なつみ
93418              岡田 ゆかり
101622             新實 さやか
139510            片山 よしあき
49636              原田 やすし
53100              菅原 あゆみ
88349        山下かおり 山下 かおり
144428            小坂田 かおり
118337             菊池 あゆみ
125519             遠藤 ひかる
85573   さんぢりあんしゃ ふぁじゃる りこ
149359              江口 まゆ
130397             戸ケ崎 りく
21395              山田 ゆかり
139445             片山 ともみ
147054             中井 あけみ
41340              釜坂 みずき
4492               吉村 きみこ
206                大野 ゆかり
66950              斉藤 あゆ美
5287               中西 みつき
121749             河野 さゆり
78982             加藤 まさ風味
22940              小島 みどり
163300             木ノ下 あき
28781              木下 ときわ
107650             多田 えりか
146998              中井 まい
103910             石原 さとみ
66806              斉藤 ひろし
82786             神田 やすたか

Now we can observe that majority of names that include Hiragana has structure of _LastName HiragaraFirstName_, with charming exceptions of weird nicknames as well as gibrish that makes no sense, here and there. 

Now onto names with Katakana

In [47]:
kata_names = names[names['name'].str.contains(katakana, regex=True)]
display(kata_names.sample(300))

name
115528                                           ワッツ 三条堺町
163205                                              木ノ下隆文
163283                                              木ノ下 忍
830                                                大野 アイリ
130393                                             戸ケ崎 泰史
158761                                           しち ゼロ ラマ
1769                                               矢ノ下 伸吾
58324                                              渡部 アサ子
162992                                             木ノ下 結衣
98547                                            アキミト アキラ
10060                                        成田 マリデル マルケズ
162984                                             木ノ下 太亮
22745                                             高根澤 ミチル
8269                                                早川 ルミ
129918                                              五ノ井 誠
163334                                              木ノ下 均
163387                                              木ノ下 博
163219                                             木ノ下 佳典
69298                                            沢田 マンション
163354                                             木ノ下 佐和
43126                                            グラフィ ソング
163028                                             木ノ下 昂生
101848                                             井ノ瀬 明弘
136980                                            更科 ファーム
62127                                              武田 ナツキ
79391                                             高田 フレディ
99930                                              新子 ミナコ
163269                                             木ノ下 理佳
9726                                              アネーロ 青森
25988                                              大塚 ミーナ
21293                                              山田 ヒロシ
163452                                             木ノ下 太郎
37214                                           石橋アンリ 真由子
58756                                              渡部 サミー
163424                                             木ノ下 信孝
66609                                              斉藤 ディナ
163037                                              木ノ下 駿
162919                                             木ノ下 帆歩
100477                                            岩崎 ナイディ
163054                                             木ノ下 竜也
30927                                             五十嵐 レイラ
12521                                           明日香 オルゴール
13176                                           リベルダーデ 山口
129271                                           石北・関口麻衣子
6469                                               中西 ケンタ
163267                                             木ノ下 繁則
89463                                            宇多津 ショット
115061                                         三条 合コンパニオン
86788                                              稲垣 メイ子
128377                                              三ツ村 聡
58300                                            渡部 ジャンペン
74430                                              後藤 ミーナ
163192                                               木ノ下勝
52205                                            浪花の ラガ男児
163060                                             木ノ下 まこ
107682                                             多田 ヒロシ
140492                                             浜田 ナオミ
92197                                              本間 ヒロシ
129931                                             五ノ井 直樹
13827                                             宮本 ジャルマ
90605                                             三浦 ノリノブ
163317                                             木ノ下 亮介
16816                                              大島 アイ子
4219                                             アンジュ パッセ
151028                                          アキヤマ カズヒロ
163488                                            木ノ下 たくみ
108657                                             滝ケ崎 富男
162945                                             木ノ下 剣斗
163365                                             

Just like the hiragana names, basic format of the names seems to be _lastName KatakanaFirstName_. 
Some exceptions are
- last name include katakana, such as 木ノ下
- fake? or made up first name in Katakana, such as ジュニア、オルゴール、etc
- non-Japanese names expressed in Katakana, e.g: グェン フ- ン
- and complete gibrish.

### Names that include ()s

In [24]:
nick_names = df[df['name'].str.contains("(", regex=False)]
print("total numbers of name records: ", df.shape[0])
print("numbers of the name records with (): ", nick_names.shape[0])
print("{} % of the records contains ().".format(round((nick_names.shape[0]/df.shape[0])*100)))
nick_names.sample(500)

total numbers of name records:  164044
numbers of the name records with ():  11619
7 % of the records contains ().


name
149575                                          米子 (Mizu)
94675                                          及川 敬子 (森川)
140311                               早戸和希 (Kazuki Hayato)
132895                               Lu Cas Mu (E Li Sha)
149083                                       江口 知佳 (浦本知佳)
131039                              加賀野 ケイ (Keiji Nakano)
90196                                          三浦 宏昭 (ミウ)
136803                              Thuỳ An (Trùm Bán Lẻ)
62738                                         田中久子 (ひさごん)
105396                            渡辺 熱 (Atsushi Watanabe)
145491                                      前川 太一 (こけし会長)
96988                                          佐野 隆 (さのり)
23585                                      小島 沙樹 (佐久間 沙樹)
162409                                   Yi-Yu Tsai (蔡易宇)
68004                                     Li Shang (ฤทัย)
74606                                        村田泉江 (みっちゃん)
68610                                         馬場完之 (馬場完之)
1362                                      加賀爪 秋夫 (ケーズビーチ)
160491                            Wei Jie Zhong (ショウ イケツ)
64648                          Cao Minh (Dao thìa dĩa gỗ)
49122                                   Lan Bàn (Tuỳ Tâm)
64585                 Dao Ngoc Toan (Quảng Cáo Thủy Toản)
135313                                 平野雅彦 (masahendrix)
128980                                          胡筠巧 (Shi)
159101                             Cheng Fu Li (Cheng Fu)
21749                                          平井 智也 (松添)
49321                                            李畹财 (En)
154876                              Ben Samai (ว่างเปล่า)
18181                                            張晶瑗 (凌殤)
161411                                         Yi Yu (浥羽)
72659                                        永勝 (Sotplug)
89166                        Cherry Nhung Võ Phạm (宇多田 光)
81819                                       佐藤 みと (谷口 みと)
84039                                        マカ 銀太 (神田洋佑)
19294                                          宮田 遼 (宮田遼)
25603                  Hua Ze Lei (Amat Bungas Penyayang)
125260                               遠藤真人 (Likalife Maco)
34728                                Sheng Gu (Christina)
14023                                      宮本直嗣 (テキーラ髭眼鏡)
15636                                   An Zhong (Madcow)
116050                                       松井 香織 (田中香織)
53010                                         菅原 直美 (なおみ)
2039                                    田辺章子 (森のキッチン アッコ)
146035                                     前川 美尋 (mihiro)
121666                            Cún Ngố (Nguyễn Bá Đạo)
88139                                        山下 大裕 (やまだい)
64617               Nguyễn Cao Nguyên (Đạo diễn cuộc đời)
83605                              神田 ちひろ (chipipipirona)
88611                                       宇多村 浩子 (Roco)
138236                                        片山真帆 (まっぴー)
157797                                         中島 麻美 (菊田)
55937                                       姫田 恵子 (宮崎 恵子)
132822                             Lu Zhuo Mu (Monica Lv)
127941                            Sonomi Kobayashi (板尾園実)
15785                                        Al An (Alan)
159272                                  LI Cheng Fu (家樂福)
26720                                      Qing Mu (Inés)
62050                                          武田 秀美 (宗田)
71052                                  Yuan Li Guo (ryan)
46557                                        山中 朝美 (花川朝美)
121243                               Đào Ngọc Toản (Bông)
13335                           宮本 真人 (Cooleyhighharmony)
129388                                    Bu Bu (xin lổi)
142057                                  Wei Jing (ainier)
14949                                          関口真奈美 (柴田)
17284                                            盧影子 (影子)
42571                                   Xi Jia-Ying (陈没话)
45749                                     浜脇 忠 (Thyouyou)
102170                                       川上 息吹 

Parenthesis mainly contains the followings
- syllabic characters (to indicate pronunciation)
- Different last names (because of marriage?)
- nick name
- Native Name in the native language, when original name is English name (or whatever language predominant in the country they're at)

many of them are valid names, so simply removing the () portion would be enough and we can treat them as the rest of the names. 

## Conclusion

While majority of the collected names are valid, there are a lot of anomalies included in the data. Best way to clean up and organize the data probably is to
1. extract names that contains the searched last name (posisble problem though, is that this might miss the different representations of last name, in Hira and Katakana and romans)
2. investigate the index of the last name in the name, to make sure the order of Last and First name. (reverse back if reversed)
3. check the presence of white space between last and first name. 
4. separate into last and first name
5. extract unique first names and put syllabic characters (use this [website](https://kanji.reader.bz/))

Then add the unique name lists of Hiragana, Kanji and Katakana to the Faker module.

Alongside of that, for data with ()s, we can simply remove the () and its contents, and put them into the pipeline above. 

Also, try using that Regex to extract Japanese letter for simpler version of the name.